### PR TITLE
mavenization and i18n

### PR DESCRIPTION
**Pull request by albfan**
standarization of code

I'm just curious about draw widget from this project, and I need to test i18n from scratch. 

Not expected you really need to merge this, but here you are.

NOTE: translations are far from finished. 
NOTE2: I choose org.m2 as package (maybe better org.2m)

### CODE REVIEW

**Comment by 2m** [comment]
Dude. I never thought I would get a PR for this project! Your are awesome! :)

Will review and merge!

File: README.md [hunk 1]
```diff
@@ -1,7 +1,6 @@
-Towers of Hanoi puzzle solver
-by Martynas Mickevičius 2010
+Towers of Hanoi puzzle solver by Martynas Mickevičius 2010
```
Comment (2m) [inline README.md:1]: You can remove my name, since it is you now as well. :) Commits tell everything about the authorship.

File: README.md [hunk 2]
```diff
@@ -12,12 +11,19 @@ to understand how recursive algorithm works.
 
 Iterative solution moves disks around in CW or CCW direction.
 
-Interface is in lithuanian language, but source code and comments
-are in english.
+Interface is i18n in english, spanish and lithuanian.
```
Comment (2m) [inline README.md:14]: You can remove word `i18n` and leave only names of the languages. And please capitalize language names.

File: src/main/resources/toh.properties [hunk 3]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Rekursyvus sprendimas. Atvaizduojamas kiekvienas pagrindin\u0117s funkcijos i\u0161kvietimas.
```
Comment (2m) [inline src/main/resources/toh.properties:1]: Not having UTF-8 in `properties` files really is not fun. Could you implement a custom `ResourceBundle.Control` [which supports UTF-8](http://stackoverflow.com/questions/4659929/how-to-use-utf-8-in-resource-properties-with-resourcebundle)?

File: src/main/resources/toh_en_US.properties [hunk 4]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Recursive solution. It shows the basic functions of each call.
+transferred.from.disk=%4d. transferred %d discs new %s ant %s\n
+solution.completed=Solution completed %s.\n
+moving.from=%d. Moving from new %s ant %s.
+step.movement=hb(%s, %s, %s, %d) '{'new %s ant %s per %s, %d disk%s'}'
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:5]: `%` should be changed to `{}`

File: src/main/resources/toh_lt_LT.properties [hunk 5]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Rekursyvus sprendimas. Atvaizduojamas kiekvienas pagrindin\u0117s funkcijos i\u0161kvietimas.
+transferred.from.disk=%4d. Perkeliamas %d diskas nuo %s ant %s\n
+solution.completed=Sprendimas baigtas per %s.\n
+moving.from=%d. Perkeliama nuo %s ant %s.
+step.movement=hb(%s, %s, %s, %d) '{'nuo %s ant %s per %s, %d disk%s'}'
```
Comment (2m) [inline src/main/resources/toh_lt_LT.properties:5]: `%` should be changed to `{}`

File: src/main/resources/toh_es_ES.properties [hunk 6]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Soluci\u00f3n recursiva. Muestra las funciones b\u00e1sicas en cada llamada.
+transferred.from.disk=%4d. Transferidos %d discos nuevo %s anterior %s\n
+solution.completed=Soluci\u00f3n completada en %s.\n
+moving.from=%d. Movimiento nuevo %s anterior %s.
+step.movement=hb({0}, {1}, {2}, {3}) '{'nuo {0} ant {1} per {2}, {3} {3,choice,1\#disco|1<discos}'}'
```
Comment (2m) [inline src/main/resources/toh_es_ES.properties:5]: Nice! I did not know about `choice` thingy. I wonder if `getLitEndingFor` could be translated to that.

File: src/main/java/org/m2/Action.java [hunk 7]
```diff
@@ -68,10 +67,10 @@ public void paintComponent(Graphics g)
 
         String text = null;
         if (diskNum != -1) {
-            text = String.format("hb(%s, %s, %s, %d) {nuo %s ant %s per %s, %d disk%s}", x.name, y.name, z.name, diskNum, x.name, y.name, z.name, diskNum, getLitEndingFor(diskNum));
+            text = MessageFormat.format(messages.getString("step.movement"), x.name, y.name, z.name, diskNum, getLitEndingFor(diskNum));
         }
         else {
-            text = String.format("%d. Perkeliama nuo %s ant %s.", TowersOfHanoi.moveCount, x.name, y.name);
+            text = String.format(messages.getString("moving.from"), TowersOfHanoi.moveCount, x.name, y.name);
```
Comment (2m) [inline src/main/java/org/m2/Action.java:73]: Should `MessageFormat.format` be used everywhere?

File: src/main/java/org/m2/Action.java [hunk 8]
```diff
@@ -1,19 +1,18 @@
+package org.m2;
```
Comment (2m) [inline src/main/java/org/m2/Action.java:1]: `org.2m` would be illegal java package name and this one does not really makes sence. Lets just change it to `toh`.

File: src/main/java/org/m2/TowersOfHanoi.java [hunk 9]
```diff
@@ -160,37 +158,37 @@ public void actionPerformed(ActionEvent ae) {
...
         p.add(interval);
         p.add(new JLabel("s."));
 
         final JTextField diskNumber = new JTextField("3", 3);
-        String[] algoTypes = { "Rekursyviai", "Iteraityviai"};
+        String[] algoTypes = {messages.getString("recursively"), messages.getString("iteratively")};
         final JComboBox algoList = new JComboBox(algoTypes);
 
-        JButton reset = new JButton("Iš naujo");
+        JButton reset = new JButton(messages.getString("New"));
```
Comment (2m) [inline src/main/java/org/m2/TowersOfHanoi.java:169]: s/New/new

File: src/main/java/org/m2/TowersOfHanoi.java [hunk 10]
```diff
@@ -160,37 +158,37 @@ public void actionPerformed(ActionEvent ae) {
...
         reset.addActionListener(new ActionListener() {
             public void actionPerformed(ActionEvent ae) {
                 try {
                     diskCount = Integer.parseInt(diskNumber.getText());
                     if (diskCount < 0 || diskCount > 100) {
-                        throw new Exception("Disk number too small.");
+                        throw new Exception(messages.getString("disk.number.too.small"));
```
Comment (2m) [inline src/main/java/org/m2/TowersOfHanoi.java:175]: I do not think exception messages should be i18nized. It does not trickle up to the GUI, does it?

File: src/main/resources/toh.properties (hunk 3 above)
Comment (2m) [inline src/main/resources/toh.properties:1]: Also we can have English as the default language.

File: src/main/resources/toh_en_US.properties [hunk 11]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Recursive solution. It shows the basic functions of each call.
+transferred.from.disk=%4d. transferred %d discs new %s ant %s\n
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:2]: Could be rephrased as `%4d. Transferred %d disc from %s to %s\n`

File: src/main/resources/toh_en_US.properties [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+solution.completed=Solution completed %s.\n
+moving.from=%d. Moving from new %s ant %s.
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:4]: `%d. Moving from %s to %s.`

File: src/main/resources/toh_en_US.properties (hunk 4 above)
Comment (2m) [inline src/main/resources/toh_en_US.properties:5]: And rephrased as `hb(%s, %s, %s, %d) '{'from %s to %s through %s, %d disk%s'}'`

File: src/main/resources/toh_en_US.properties [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+step.movement=hb(%s, %s, %s, %d) '{'new %s ant %s per %s, %d disk%s'}'
+disk.moved.counterclockwise.direction=Even number of discs - Movement in counterclockwise direction.
+disk.moved.in.clockwise.direction=Odd number of disc - disc moved in clockwise direction.
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:7]: s/disc/discs

File: src/main/resources/toh_en_US.properties [hunk 14]
```diff
@@ -0,0 +1,27 @@
+recursive.solution=Recursive solution. It shows the basic functions of each call.
+transferred.from.disk=%4d. transferred %d discs new %s ant %s\n
+solution.completed=Solution completed %s.\n
+moving.from=%d. Moving from new %s ant %s.
+step.movement=hb(%s, %s, %s, %d) '{'new %s ant %s per %s, %d disk%s'}'
+disk.moved.counterclockwise.direction=Even number of discs - Movement in counterclockwise direction.
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:6]: s/Movement/disc moved

File: src/main/resources/toh_en_US.properties [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+disk.moved.in.clockwise.direction=Odd number of disc - disc moved in clockwise direction.
+iterative.solution=Iterative solution
+look.and.feel.was.not.found=Look and feel was not found.
+towers.of.hanoi=Towers of Hanoi
+resolved=resolved
+initial.number.of.discs=Initial number of disc\:
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:12]: s/disc/discs

File: src/main/resources/toh_en_US.properties [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+incorrect.number.of.discs=Incorrect number of discs
+number.of.disc.must.be.between.0.and.100=Number of disc must be between 0 and 100
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:14]: s/disc/discs

File: src/main/resources/toh_en_US.properties [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+continue=continue
+New=New
+iteratively=iteratively
+recursively=recursively
+step.is=Step is\:
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:19]: s/is/takes

File: src/main/resources/toh_en_US.properties [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+improper.pitch.range=Improper pitch range
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:20]: `Invalid step duration`

File: src/main/resources/toh_en_US.properties [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+step.interval.must.be.greater.than.or.equal.zero=Step interval must be greater than or equal zero.
```
Comment (2m) [inline src/main/resources/toh_en_US.properties:21]: s/interval/duration

**Comment by 2m** [comment]
Good stuff! If you can address these various nitpicks, then I will merge this. Just drop a note here when you push some more commits.

Again, thanks for contributing back! :)

**Comment by 2m** [comment]
Hey. Sorry for spaming the PR with a storm of comments. Today I have stumbled upon this talk https://www.youtube.com/watch?v=O8CbzKREAj4 by late Pieter Hintjens. And one of the most touching ideas of his was "merge first, review later". Therefore I came back here to do just that (albeit four years late).

Thanks again! :)

**Comment by albfan** [comment]
Nice, but I think you're right in first case.

What you can do is to create a wip/branch with all the fixes and split PR into smaller pieces. Whenever something is in shape merge it (or cherry-pick)

My choose: undo the merge and let's fix all that review, I'm sure I can't fix almost everything you suggest

Anyway, now it is on maven hack on it would be easy

**Comment by 2m** [comment]
I fixed up almost everything I mentioned here in a subsequent PR https://github.com/2m/toh/pull/2 :)

**Comment by albfan** [comment]
Great then!

**Pull request merged**